certchain/certmanager/helpers_test.go

```diff
@@ -93,15 +94,31 @@ func (p *stubProducer) Stop() { close(p.out) }
 var _ certmanager.Cache = (*stubCache)(nil)
 
 type stubCache struct {
-	OnWrite chan *certchain.AugmentedChain
+	avail    chan struct{}
+	chainMap map[string]*certchain.AugmentedChain
 }
 
 func newStubCache() *stubCache {
-	return &stubCache{make(chan *certchain.AugmentedChain, 1)}
+	return &stubCache{
+		make(chan struct{}, 1),
+		make(map[string]*certchain.AugmentedChain),
+	}
+}
+
+func (c *stubCache) Read(digest string) (*certchain.AugmentedChain, error) {
+	<-c.avail
```

Comment:
The data is removed from the channel once it's read, and so the channel becomes empty. It means the second `Read` operation after a `Write` (or any `Read` operation without `Write`) incurs reading from empty channel, which blocks the execution forever.

* Apply a timeout whenever reading from a channel. Tests not finishing forever are harmful, even if they're merely caused by wrong code.

* Exclude the channel read from this method. This behavior is really confusing: `Read` is not assumed to wait for `Write`. I already made a wrong comment. If you hate to write the channel read in the test code, consider adding a new method.

Reply:
Done